### DISCUSSION TITLE
Incremental Writing of Compressed Multi-HDU Images

### DISCUSSION THREAD

**Discussion by keastrid** [nom-tam-fits/nom-tam-fits#334] [Q&A] [answered]
Is there a way to incrementally write multiple compressed image HDUs to the same file?

```java
FitsOutputStream out = ...;

for (...) {
  ImageHDU hdu = ...;
  ...
  CompressedImageHDU cHDU = ...;
  cHDU.compress();
  hdu.write(out);
}
out.close();
```

**Comment by attipaci** [comment, marked as answer]
I think the above should almost work. The incremental writing (as of 1.17.0) does not care whether the HDU is compressed or not. The only thing that might stand in the way is that the primary HDU cannot be a compressed one since compressed HDUs are essentially binary tables, which may only be extensions. So, you need to make a dummy hdu (header only) as the primary, and then you can add the compressed HDUs after. Something like:

```java
FitsOutputStream out = ...;

// Write a primary HDU with the minimum required header
new ImageHDU(new Header(), null).write(out);

// Now add the compressed image HDUs as extensions...
for (...) {
  ImageHDU hdu = ...;
  ...
  CompressedImageHDU cHDU = ...;
  cHDU.compress();
  hdu.write(out);
}

out.close();
```

should work. 

Let me know how it goes.

PS. Somewhat relatedly, I created issue #335 for making the creation of dummy (or header-only) HDUs more intuitive. I plan to add that feature in 1.18...

**Reply by keastrid** [reply]
This seems to work - have to adjust my reader to handle it, but the writing seems to be working well.

Thanks!
